### DISCUSSION TITLE
Understanding the difference between invoking cy commands inside a promise in these two examples

### DISCUSSION THREAD

**Discussion by zakariaelas** [cypress-io/cypress#25975] [Questions and Help]
Hello!

Would appreciate anyone helping me wrap my head around the following two examples:

### Example 1

Initially, I did something like this:

```javascript
Cypress.Commands.add('customCommand', async () => {
  const check = await checkAsyncThing(region);

  if (check) {
    await doAnotherThing();
    cy.log(`did another thing`);
  }
});
```

Then I invoked the custom command from my test.

From the above example, I got this error:

> Cypress detected that you returned a promise from a command while also invoking one or more cy commands in that promise.

### Example 2

I tried fixing the above error by doing the following:

```javascript
const asyncWork = async () => {
  const check = await checkAsyncThing(region);

  if(check) {
    await doAnotherThing()
    cy.log(`did another thing`);
  }
};

Cypress.Commands.add('customCommand', () => {
  cy.wrap(null).then(() => asyncWork());
});
```

This worked, but I don't really understand why. I can see that in Example 2 my custom command is no longer returning a promise (removal of `async` from the arrow function), however, `then` is also technically returning a promise in which I'm calling a cy command (`cy.log`). 

Can you help me understand why Example 2 works, but not Example 1? Would really appreciate help building the right mental model for working with Cypress.

Thanks!

**Comment by portyu9** [comment]
Hi @zakariaelas 

The important distinction is that Cypress commands are **not native Promises**.

In Example 1 your custom command itself is `async`, so it immediately returns a native Promise. While that Promise is executing, it also queues `cy.log()`. Cypress rejects that combination because it now has two different schedulers to reconcile:

- the native Promise/microtask queue
- Cypress's own serial command queue

In Example 2, `then()` is **Cypress's `.then()` command**, not `Promise.prototype.then()`. Cypress explicitly supports a Promise being returned from a `.then()` callback and waits for it before advancing the command queue.

I would still avoid putting Cypress commands such as `cy.log()` inside the native async function. Keep the Promise-only work separate and return to Cypress's chain afterward:

```js
const asyncWork = async () => {
  const check = await checkAsyncThing(region)

  if (!check) {
    return false
  }

  await doAnotherThing()
  return true
}

Cypress.Commands.add('customCommand', () => {
  return cy
    .then(() => asyncWork())
    .then((didWork) => {
      if (didWork) {
        cy.log('did another thing')
      }
    })
})
```

A useful mental model is: **use native `await` inside native async code, and use Cypress chaining for Cypress commands. Cross between the two at a Cypress `.then()` boundary.**

Please mark this answer as accepted if it helped, thank you.
